### PR TITLE
Quadicon implementation in React

### PR DESCRIPTION
**Pull request by martinpovolny**
Mostly as an experiment to see how much work it would be to reimplement the Angular components that we have in React, I am adding an implementation or Quadicons here.

Working on that I found some issues that I am also fixing. This PR is a temporary place for all the fixes and a place to discuss. I am ready to make the minor issues into separate PRs or address the issues in a different way if needed.

ping @karelhala, @himdel, @skateman

### CODE REVIEW

**Comment by martinpovolny** [comment]
After: 
quadicons-strory

**Comment by martinpovolny** [comment]
Ping @epwinchell @terezanovotna : can you, please, check this?

The styling is taken from the Angular implementation so there should be no change.

**Comment by skateman** [comment]
@martinpovolny I'm curious about the tooltips, do they work?

**Comment by terezanovotna** [comment]
There is a huge opportunity for some design improvements. When refactored, happy to work on making it more usable and nicer. 

At this moment, the interaction with quadicons is very confusing.

File: src/quadicon/Quaditem.jsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        {shortText}
+      </div>));
+  }
+  if (props.piechart || props.piechart === 0) {
+    parts.push((
+      <div className={`piechart fill-${props.piechart}`} />
+    ));
+  }
+  return (
+    <div className={`miq-quaditem ${props.cls}`} title={props.tooltip} style={{ background: props.background }}>
```
Comment (karelhala) [inline src/quadicon/Quaditem.jsx]: Do not use `cls` but rather `className`. Also if `cls` is undefined you shouldn't add it at all - 
```
<div className={`miq-quaditem ${cassName ? className : ''}`}
```

Comment (himdel) [reply src/quadicon/Quaditem.jsx]: maybe `cx('miq-quaditem', props.cls)` ?

**Comment by epwinchell** [comment]
@martinpovolny Nice. Looks identical to the Angular Implementation.

**Comment by epwinchell** [comment]
> At this moment, the interaction with quadicons is very confusing.

@terezanovotna I'm happy to work with you. It probably makes sense to do that in follow-up PRs.

**Comment by martinpovolny** [comment]
> There is a huge opportunity for some design improvements. When refactored, happy to work on making it more usable and nicer.

Well I am really aiming just to migrate the code from Angular to React is this PR.

File: src/quadicon/Quaditem.jsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    return 'font-normal';
+  } else if (len > 2 && len < 4) {
+    return 'font-small';
+  }
+  return 'font-tiny';
+};
+
+const Quaditem = (props) => {
+  const shortText = abbrNumber(props.text);
+  const parts = [];
```
[the remark this answers is not in the document]
Comment (martinpovolny) [reply src/quadicon/Quaditem.jsx]: ok, can be

File: src/quadicon/Quaditem.jsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  if (props.fonticon) {
+    parts.push((
+      <div className="fonticon">
```
[the remark this answers is not in the document]
Comment (martinpovolny) [reply src/quadicon/Quaditem.jsx]: Afaik there are existing classes, I cannot change it w/o changing ManageIQ UI. Those are just copied to the notebook to make it work.

File: src/quadicon/Quaditem.jsx (hunk 1 above)
Comment (martinpovolny) [reply src/quadicon/Quaditem.jsx]: ok

File: src/quadicon/Quadicon.jsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import Quaditem from './Quaditem';
+import './styles.scss';
+
+const quadSet = ['topLeft', 'topRight', 'bottomLeft', 'bottomRight', 'middle'];
+
+// const getBackground = (data, item) => (
+//   data[item] && data[item].background ? { background: data[item].background } : {}
+// ); // FIXME: check if we can just pass an undefined background in the markup
+
+const renderSingle = item => (
```
[the remark this answers is not in the document]
Comment (martinpovolny) [reply src/quadicon/Quadicon.jsx]: ok

File: src/quadicon/Quadicon.jsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      text={item.text}
+      tooltip={item.tooltip}
+      background={item.background}
+      color={item.color}
+      piechart={item.piechart}
+    />
+  </div>
+);
+
+const renderQuad = data => (
```
[the remark this answers is not in the document]
Comment (martinpovolny) [reply src/quadicon/Quadicon.jsx]: ok

**Comment by martinpovolny** [comment]
@himdel, @karelhala : I resolved most of the review feedback.

Also fixing broken tests on master.

File: README.md [hunk 6]
```diff
@@ -70,19 +70,23 @@ It will automatically watch your files and store final files into `dist` folder.
...
 your components and you can debug them in node server environment.
 
+The server will run at http://localhost:8080.
+
 #### 3. `[npm|yarn] run storybook`
 It will automatically watch your files and presents them using [storybook](https://storybook.js.org/), if you want
 to use your new components, edit stories inside [stories/index.stories.js](/stories/index.stories.js) and debug them in
 nice storybook UI.
 
+The storybook server will run ad http://localhost:6006.
```
Comment (skateman) [inline README.md]:
```suggestion
The storybook server will run at http://localhost:6006.
```

File: src/quadicon/Quadicon.jsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import kebabCase from 'lodash/kebabCase';
+
+import Quaditem from './Quaditem';
+import './styles.scss';
+
+const quadSet = ['topLeft', 'topRight', 'bottomLeft', 'bottomRight', 'middle'];
+
+// const getBackground = (data, item) => (
+//   data[item] && data[item].background ? { background: data[item].background } : {}
+// ); // FIXME: check if we can just pass an undefined background in the markup
```
Comment (skateman) [inline src/quadicon/Quadicon.jsx]: Is this still a FIXME?

File: README.md (hunk 6 above)
Comment (martinpovolny) [reply README.md]: fixed

**Comment by martinpovolny** [comment]
I fixed both @skateman 's comments. Thx for the review!

File: src/vendor.scss [hunk 8]
```diff
@@ -9,6 +9,6 @@
 */
 @import '~patternfly-react/dist/css/patternfly-react.css';
 
-@import "~react-select/dist/react-select.css";
+/* @import "~react-select/dist/react-select.css"; */
```
Comment (himdel) [inline src/vendor.scss]: Is this relevant?
Doesn't seem related to quadicons..

Comment (martinpovolny) [reply src/vendor.scss]: Not related to quadicons as are not several other commits above.

Only needed to make the stuff actuall work as whole.

Do you want me to separate the unrelated stuff (already in separate commits) to a separate PR?

Comment (himdel) [reply src/vendor.scss]: Nope, just wanted to make sure this is intentional and fixing things, not breaking other things :).
(More like, sounds it could break some selects?)

But I see the commit message mentions it's missing, so, probably not breaking anything.

Comment (martinpovolny) [reply src/vendor.scss]: I did not know about his part do I just commented this out. I hope someone who knows about the selects can take a look in a separate PR.

However for now w/o commenting this out, the package would not build.

**Comment by himdel** [comment]
LGTM, but also I don't see a point of *not* merging because current master can't be built and won't pass tests :).

This fixes the build part.
